### PR TITLE
min gas price for block producers

### PR DESCRIPTION
**Pull request by tkstanczak**
Allows setting target gas block limit
Adds config migrations
Renames abstract classes so the 'base' word is at the end which is the convention
Move ITxFilter to Consensus and implements MinGasPriceFilter
Removes LruCache and replaces it permanently to LruCacheWithRecycling

### CODE REVIEW

File: src/Nethermind/Nethermind.Blockchain/BlockTree.cs [hunk 1]
```diff
@@ -43,8 +43,8 @@ public partial class BlockTree : IBlockTree
     {
         private const long LowestInsertedBodyNumberDbEntryAddress = 0; 
         private const int CacheSize = 64;
-        private readonly ICache<Keccak, Block> _blockCache = new LruCacheWithRecycling<Keccak, Block>(CacheSize, CacheSize, "blocks");
-        private readonly ICache<Keccak, BlockHeader> _headerCache = new LruCacheWithRecycling<Keccak, BlockHeader>(CacheSize, CacheSize, "headers");
+        private readonly ICache<Keccak, Block> _blockCache = new LruCache<Keccak, Block>(CacheSize, CacheSize, "blocks");
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/BlockTree.cs:46]: why not recycling? I Lru cache now recycling?

File: src/Nethermind/Nethermind.Blockchain/Producers/BlockProducerBase.cs [hunk 2]
```diff
@@ -187,8 +194,24 @@ protected virtual Block PrepareBlock(BlockHeader parent)
...
+            {
+                IReleaseSpec spec = _specProvider.GetSpec(parent.Number + 1);
+                long maxGasLimitDifference = parentGasLimit / spec.GasLimitBoundDivisor;
+                gasLimit = targetGasLimit.Value > parentGasLimit
+                    ? parentGasLimit + Math.Min(targetGasLimit.Value - parentGasLimit, maxGasLimitDifference)
+                    : parentGasLimit - Math.Min(parentGasLimit - targetGasLimit.Value, maxGasLimitDifference);
+            }
+
+            return gasLimit;
+        }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/Producers/BlockProducerBase.cs:213]: I would put it behind some IGasLimitCalculator, make GasLimitOverride in AuRa just a type of decorator over this calculator

File: src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs [hunk 3]
```diff
@@ -42,7 +42,7 @@ public class PersistentReceiptStorage : IReceiptStorage
         private static readonly ReceiptStorageDecoder StorageDecoder = ReceiptStorageDecoder.Instance;
         
         private const int CacheSize = 64;
-        private readonly ICache<Keccak, TxReceipt[]> _receiptsCache = new LruCacheWithRecycling<Keccak, TxReceipt[]>(CacheSize, CacheSize, "receipts");
+        private readonly ICache<Keccak, TxReceipt[]> _receiptsCache = new LruCache<Keccak, TxReceipt[]>(CacheSize, CacheSize, "receipts");
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs:45]: ?

File: src/Nethermind/Nethermind.Consensus/Transactions/FilteredTxSource.cs [hunk 4]
```diff
@@ -16,27 +16,32 @@
...
-        public override string ToString() => $"{nameof(TxFilterTxSource)} [ {_innerSource} ]";
+        public IEnumerable<Transaction> GetTransactions(BlockHeader parent, long gasLimit)
+        {
+            foreach (Transaction transaction in _innerSource.GetTransactions(parent, gasLimit))
+            {
+                if (_txFilter.IsAllowed(transaction, parent))
+                {
+                    yield return transaction;
+                }
+            }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Transactions/FilteredTxSource.cs:42]: Same code in 7 lines

File: src/Nethermind/Nethermind.Blockchain/BlockTree.cs (hunk 1 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Blockchain/BlockTree.cs:46]: yes, LruCache was not used anywhere so I renamed recycling to normal and removed the other one

File: src/Nethermind/Nethermind.Blockchain/Producers/BlockProducerBase.cs (hunk 2 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Blockchain/Producers/BlockProducerBase.cs:213]: not touching the AuRa gas limit override, not needed here, also mine approach is simpler than introducing other classes

File: src/Nethermind/Nethermind.Consensus/Transactions/FilteredTxSource.cs (hunk 4 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Consensus/Transactions/FilteredTxSource.cs:42]: but does not allocate any delegates on each block and follows the rule of not using LINQ unless absolutely necessary or beneficial

File: src/Nethermind/Nethermind.Consensus.AuRa/AuRaContractGasLimitCalculator.cs [hunk 5]
```diff
@@ -21,13 +21,12 @@
...
 using Nethermind.Core.Caching;
 using Nethermind.Core.Crypto;
-using Nethermind.Crypto;
 using Nethermind.Dirichlet.Numerics;
 using Nethermind.Logging;
 
 namespace Nethermind.Consensus.AuRa
 {
-    public class AuRaContractGasLimitOverride : IGasLimitOverride
+    public class AuRaContractGasLimitCalculator : IGasLimitCalculator
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus.AuRa/AuRaContractGasLimitCalculator.cs:29]: can we call it AuRaContractGasLimitOverrideCalculator ? ContractGasLimitOverride is kind og domain name

File: src/Nethermind/Nethermind.Consensus/GasLimitCalculator.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+// 
+
+using System;
+using Nethermind.Core;
+using Nethermind.Core.Specs;
+
+namespace Nethermind.Consensus
+{
+    public class GasLimitCalculator : IGasLimitCalculator
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/GasLimitCalculator.cs:24]: TargetBlockGasLimitCalculator?

File: src/Nethermind/Nethermind.Consensus.AuRa/AuRaContractGasLimitCalculator.cs (hunk 5 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Consensus.AuRa/AuRaContractGasLimitCalculator.cs]: done, thanks

File: src/Nethermind/Nethermind.Consensus/GasLimitCalculator.cs (hunk 6 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Consensus/GasLimitCalculator.cs]: TargetAdjusted

**Pull request merged**